telegram/telegram.py

```diff
@@ -0,0 +1,199 @@
+            log.info('Inbound update does not contain a message')
+            request.finish()
+            return
+        else:
+            message = update['message']
+
+        if 'text' in message:
+            message = self.translate_inbound_message(update['message'])
+            log.info(
+                'TelegramTransport receiving inbound message from %s to %s' % (
+                    message['from_addr'], message['to_addr']))
+
+            yield self.publish_message(
+                message_id=message_id,
+                content=message['content'],
+                to_addr=message['to_addr'],
+                from_addr=message['from_addr'],
+                transport_type=self.transport_type,
+                transport_name=self.transport_name,
+            )
+        else:
+            log.info('Message is not a text message')
+
+        request.finish()
+
+    def translate_inbound_message(self, message):
+        """
+        Translates inbound Telegram message into Vumi's preferred format
+        """
+        content = message['text']
+        to_addr = self.bot_username
+        # Sender field is empty if message is sent over a Telegram channel as
+        # opposed to being directly sent to our bot (in which case the channel
+        # id is what we want)
+        if 'from' in message:
+            from_addr = message['from']['id']
+        else:
+            from_addr = message['chat']['id']
+
+        message = {
+            'content': content,
+            'to_addr': to_addr,
+            'from_addr': from_addr,
+        }
+        return message
+
+    @inlineCallbacks
+    def handle_outbound_message(self, message):
+        # TODO: handle direct replies
+        message_id = message['message_id']
+        outbound_msg = {
+            'chat_id': message['to_addr'],
+            'text': message['content'],
+        }
+        url = '%s/sendMessage' % self.outbound_url.rstrip('/')
+        http_client = HTTPClient(self.agent_factory())
+
+        try:
+            r = yield http_client.post(
+                url=url,
+                data=json.dumps(outbound_msg),
+                headers={'Content-Type': ['application/json']},
+            )
+            content = yield r.content()
+            res = json.loads(content)
+            if r.code == http.OK and res['ok']:
+                yield self.publish_ack(
+                    user_message_id=message_id,
+                    sent_message_id=message_id,
+                )
+            else:
+                yield self.outbound_failure(message_id, res['description'])
+
+        # Treat page redirects as errors, since Telegram seems to redirect us
+        # when our bot token is invalid
+        except ResponseFailed:
+            yield self.outbound_failure(message_id, 'Invalid token')
+        # In case we get a response from Telegram that isn't JSON
+        except ValueError:
+            yield self.outbound_failure(message_id, 'Unexpected response')
```

Comment:
Generally it's useful to try and include the error message from the transport in the reason. That way you can provide more context to an app developer using the transport, things like throttling or invalid tokens etc... are useful to receive as reasons.